### PR TITLE
4.x: Mocking documentation

### LINKED ISSUE #8339: 4.x: Document CDI based mocking for unit tests
## Environment Details
* Helidon Version: 4.0.4
* Helidon MP
----------

## Problem Description

Mockito's `@Mock` does not work with `@HelidonTest`.

---

Consider the following JAXRS resource and CDI bean:

```java
@ApplicationScoped
public class FooService {

    public String getFoo() {
        return "foo";
    }
}

@Path("/foo")
public class FooResource {

    @Inject
    private FooService fooService;

    @GET
    public String getFoo() {
        return fooService.getFoo();
    }
}
```

---

A test that combines both `@HelidonTest` and `@Mock` does not work:

```java
@HelidonTest
@ExtendWith(MockitoExtension.class)
class AcmeTest {
 
    @Inject
    private WebTarget target;

    @Mock
    private AcmeService acmeService;

    @Test
    void testFoo() {
        when(acmeService.getFoo()).thenReturn("bar");

        Response response = target.path("/foo").request().get();

        assertThat(response.getStatus(), is(200));
        assertThat(response.readEntity(String.class), is("bar"));
     }
}
```

----

### Current solution

Instead, we need to use CDI to provide an alternative. This solution works with Helidon 2.x, 3.x and 4.x.

```java
@HelidonTest
@Priority(1) // required for @Alternative
class FooTest {

    @Inject
    private WebTarget target;

    private FooService fooService;

    @BeforeEach
    void initMock() {
        // set fooService to a new mock before each test
        // use real methods for default answers
        fooService = Mockito.mock(FooService.class, InvocationOnMock::callRealMethod);
    }

    @Produces
    @Alternative
    FooService mockFooService() {
        // This makes FooResource inject our mock instead of the default singleton
        return fooService;
    }

    @Test
    void testMockedFoo() {
        // test the modified behavior
        when(fooService.getFoo()).thenReturn("bar");

        Response response = target.path("/foo").request().get();

        assertThat(response.getStatus(), is(200));
        assertThat(response.readEntity(String.class), is("bar"));
    }

    @Test
    void testFoo() {
        // test the non modified behavior
        Response response = target.path("/foo").request().get();

        assertThat(response.getStatus(), is(200));
        assertThat(response.readEntity(String.class), is("foo"));
    }
}
```

pom.xml:
```xml
<dependency>
    <groupId>org.mockito</groupId>
    <artifactId>mockito-core</artifactId>
    <version>5.8.0</version>
    <scope>test</scope>
</dependency>
```

---

Action items:
- [ ] document how to setup mocks using CDI producer methods (I.e. the example above)
- [ ] Investigate if there is a way to make `HelidonTest` and `MockitoExtension` work together.
- [ ] Investigate if we can provide an annotation to setup a mock as an alternative
- [ ] Investigate if we should support other use-cases (e.g. `@org.mockito.Spy`, `@org.mockito.Captor`)

Example of a custom annotation to polish the example above:

```java
@HelidonTest
class FooTest {

    // Creates a new mock instance for each test
    // and expose it as an alternative
    @io.helidon.microprofile.testing.junit5.Mock
    private FooService fooService;

    @Test
    void testMockedFoo() {
        // same as above
    }
}
```

### LINKED ISSUE #8772: 4.x: Document @MockBean unit tests
[//]: # "Please remove these comments"
[//]: # "Provide a general summary of the issue in the Title above"

## Environment Details
* Helidon Version: 4
* Helidon MP
* JDK version: 21

----------

## Problem Description
[//]: # "Describe the bug in detail highlighting current behavior vs expected behavior"
We have to document this PR that contains a new annotation @MockBean and GAV io.helidon.microprofile.testing:helidon-microprofile-testing-mocking:
https://github.com/helidon-io/helidon/pull/8674

### PR DESCRIPTION
**Pull request by tvallin**
### Description

Add documentation about CDI based mocking documentation:
 - Introduce new doc package named `testing` that gather testing snippet and related `*.adoc`.
 - New code snippets.
 - New mock section in `testing.adoc`.

fix #8339
fix #8772

### Documentation

Updated into this PR

### CODE REVIEW

File: docs/src/main/asciidoc/mp/testing.adoc [hunk 1]
```diff
@@ -158,6 +159,36 @@ include::{sourcedir}/mp/TestingSnippets.java[tag=snippet_2, indent=0]
 <3> Add JaxRs support to the current test class.
 <4> Define a `RequestScoped` bean.
 
+== Mock Support
+
+One way of using Mocks with Helidon testing framework is to use CDI. The following example shows how to mock a service
```
Comment (romain-grecourt) [inline docs/src/main/asciidoc/mp/testing.adoc:164]: CDI can be used to enable mocking, the following example shows

**Review by romain-grecourt — COMMENTED** [review]
How does this fit with #8772 ?

**Comment by tvallin** [comment]
I see it as 2 chapters under `Mock Support` where the first one would be helidon mocking support and the second mocking support with CDI.

**Comment by romain-grecourt** [comment]
> I see it as 2 chapters under `Mock Support` where the first one would be helidon mocking support and the second mocking support with CDI.

Can you reflect this in your changes ?
I.e. the section title and the snippet class name.

**Review by romain-grecourt — APPROVED** [review]

**Pull request merged**